docker/runtime/nodejs/kubeless.js

```diff
@@ -18,89 +23,114 @@ const funcHandler = process.env.FUNC_HANDLER;
-        _res.end(result);
-        break;
-      case 'object':
-        _res.json(result);
-        break;
-      default:
-        _res.end(JSON.stringify(result))
-    }
-    _end();
-  }).catch((err) => {
-    // Catch asynchronous errors
-    _handleError(err);
-  });
-} catch (err) {
-  // Catch synchronous errors
-  _handleError(err);
-}`;
-const { vmscript, sandbox } = helper.loadFunc(modName, functionCallingCode);
 
-app.all('*', (req, res) => {
-  res.header('Access-Control-Allow-Origin', '*');
-  if (req.method === 'OPTIONS') {
-    // CORS preflight support (Allow any method or header requested)
-    res.header('Access-Control-Allow-Methods', req.headers['access-control-request-method']);
-    res.header('Access-Control-Allow-Headers', req.headers['access-control-request-headers'])
-    res.end();
-  } else {
-    const funcLabel = modName + "-" + req.method;
-    const end = statistics.timeHistogram.labels(funcLabel).startTimer();
-    statistics.callsCounter.labels(funcLabel).inc();
-    const handleError = (err) => {
-      statistics.errorsCounter.labels(funcLabel).inc();
-      res.status(500).send('Internal Server Error');
-      console.error(`Function failed to execute: ${err.stack}`);
-    };
-    let data = req.body.toString('utf-8');
-    if (req.get('content-type') === 'application/json') {
-      data = JSON.parse(data)
-    }
-    const reqSandbox = Object.assign({
-      event: {
-        data,
-        'event-type': req.get('event-type'),
-        'event-id': req.get('event-id'),
-        'event-time': req.get('event-time'),
-        'event-namespace': req.get('event-namespace'),
-        extensions: {
-          request: req,
-        },
-      },
-      context: {
-        'function-name': funcHandler,
-        timeout,
-        runtime: process.env.FUNC_RUNTIME,
-        'memory-limit': process.env.FUNC_MEMORY_LIMIT,
-      },
-      _handleError: handleError,
-      _res: res,
-      _end: end,
-      process: Object.assign({}, process),
-    }, sandbox);
+const script = new vm.Script(fs.readFileSync(modPath), {
+    filename: modPath,
+    displayErrors: true,
+});
+
+function modRequire(p, req, res, end) {
+    if (p.indexOf('./') === 0)
+        return require(path.join(path.dirname(modPath), p));
+    else if (libDeps.includes(p))
+        return require(path.join(libPath, p));
+    else if (p === 'kubeless')
+        return (handler) => modExecute(handler, req, res, end);
+    else
+        return require(p);
+}
+
+function modExecute(handler, req, res, end) {
     try {
-      vmscript.runInNewContext(reqSandbox, { timeout: timeout*1000 });
+        const event = {
+            'event-type': req.get('event-type'),
+            'event-id': req.get('event-id'),
+            'event-time': req.get('event-time'),
+            'event-namespace': req.get('event-namespace'),
+            'data': req.body.toString('utf-8'),
+            'extensions': { request: req },
+        };
+        const context = {
```

Comment:
I would move the context initialization out of the modExecute function so it is executed just once (instead of one per request). That way we can save some ms.

Reply:
Yes. And at this point we could also check the Promise callbacks. Instead of `(err) => { handleError(err); }` just `err => handleError(err, ...)` at least under node 6 a little faster to my knowledge. Of course also the typical case, the success with then `result => handleResult(result)` using an additional plain function.